### ISSUE TITLE
Allow throwing functions to be curried

### ISSUE THREAD

**Issue by elliottwilliams** [thoughtbot/Curry#34]
Is there any reason why this hasn't been done? Seems like it would just require adding `throws` to the generated curry functions.

Usage would be like:

``` swift
func foo(a: Int, b: Int) throws -> Int { ... }

let curried = curry(foo)
// curried: (a: Int) -> (b: Int) throws -> Int

try curried(1)(2)

```

**Comment by gfontenot** [comment]
Pretty sure I tried this at one point and it wasn't possible. If you can get it to compile I'd be happy to look at a PR.

**Comment by sharplet** [comment]
Here's what I imagine the definition of a throwing 2-argument curry to be:

``` swift
func curry<A, B, C>(f: (A, B) throws -> C) -> A -> B rethrows -> C {
  return { a in { b in { try f(a, b) } } }
}
```

But it doesn't compile:

```
error: repl.swift:1:54: error: only function declarations may be marked 'rethrows'
func curry<A, B, C>(f: (A, B) throws -> C) -> A -> B rethrows -> C {
                                                     ^
```

Without `rethrows` all uses of `curry` would need to be marked with `try`, so I'm not sure this is possible?

**Comment by elliottwilliams** [comment]
You're right. I'd be interested in better understanding why that error exists, but for now, I can get it working by overloading each `curry` function to have a throwing and non-throwing version:

``` swift
func curry<A, B>(function: (A) -> B) -> (A) -> B {
    return { (`a`: A) -> B in function(`a`) }
}
func curry<A, B>(function: (A) throws -> B) -> (A) throws -> B {
    return { (`a`: A) throws -> B in try function(`a`) }
}

func curry<A, B, C>(function: (A, B) -> C) -> (A) -> (B) -> C {
    return { (`a`: A) -> (B) -> C in { (`b`: B) -> C in function(`a`, `b`) } }
}
func curry<A, B, C>(function: (A, B) -> C) -> (A) -> (B) throws -> C {
    return { (`a`: A) -> (B) throws -> C in { (`b`: B) throws -> C in function(`a`, `b`) } }
}
// and so on...
```

I'd be happy to put together a PR that does this, but I understand if you'd rather not double the number of functions being generated.

**Comment by gfontenot** [comment]
I definitely don't want to double the number of functions. That would seriously limit the number of implementations we could reasonably ship (we're already limited by exponential compile times). My assumption is that this would add an unreasonable amount of complexity for the compiler.

I'm not sure what the actual answer here is.

**Comment by elliottwilliams** [comment]
I tried rewriting the functions using the curry syntax being removed in Swift 3:

``` swift
func curry<A, B>(function: (A) throws -> B)(_ a: A) rethrows -> B {
    return try function(a)
}

func curry<A, B, C>(function: (A, B) throws -> C)(_ a: A)(_ b: B) rethrows -> C {
    return try function(a, b)
}
```

These compile, but `rethrows` doesn't seem to be working properly, i.e. all types returned from these functions are throwing. In testing, I came across this:

``` swift
func a2(x: Int) -> Int {
    return x+2
}
let ca2: Int -> Int = curry(a2)
// Error: 'curry' produces '(A) -> (B) throws -> C', not the expected contextual result type 'Int -> Int'
```

Given that error message, I'm guessing that under the old currying syntax, this edge case was handled by turning `rethrows` into `throws`, and there is (still) no way to return a rethrowing closure.

At the moment I'm looking to see if this limitation has been discussed before, or if I should file a Swift bug. Regardless, seems like the current answer to my original question is "it's impossible without further overloading" :(
